File: examples/wasm/grayskull.c
#include <stddef.h>

// A simple bump allocator for our WASM module since we don't have stdlib.
#define MEMORY_HEAP_SIZE (1024 * 1024 * 8)  // 8MB heap for a few buffers
static unsigned char memory_heap[MEMORY_HEAP_SIZE];
static size_t heap_ptr = 0;
void* gs_alloc(size_t size) {
  size_t aligned_size = (size + 7) & ~7;
  if (heap_ptr + aligned_size > MEMORY_HEAP_SIZE) return NULL;
  void* ptr = &memory_heap[heap_ptr];
  heap_ptr += aligned_size;
  return ptr;
}
void gs_free(void* ptr) { (void)ptr; /* No-op for bump allocator */ }
void gs_reset_allocator(void) { heap_ptr = 0; }

// Minimal standard library functions for WASM nostdlib build
void* memset(void* s, int c, size_t n) {
  unsigned char* p = (unsigned char*)s;
  for (size_t i = 0; i < n; i++) p[i] = (unsigned char)c;
  return s;
}

void* memcpy(void* dest, const void* src, size_t n) {
  unsigned char* d = (unsigned char*)dest;
  const unsigned char* s = (const unsigned char*)src;
  for (size_t i = 0; i < n; i++) d[i] = s[i];
  return dest;
}

#define gs_assert(cond)
#define GS_NO_STDLIB
#define GS_API
#include "../../grayskull.h"
#include "../nanomagick/frontalface.h"

#define NUM_BUFFERS 3
static struct gs_image images[NUM_BUFFERS];

// Functions to be exported to WASM
void gs_init_image(int idx, int w, int h) {
  if (idx < 0 || idx >= NUM_BUFFERS) return;
  // Simple bump allocation, so we just allocate once on first use.
  // gs_reset_allocator() must be called from JS if sizes change.
  if (images[idx].data == NULL) { images[idx].data = gs_alloc(w * h); }
  images[idx].w = w;
  images[idx].h = h;
}

uint8_t* gs_get_image_data(int idx) {
  if (idx < 0 || idx >= NUM_BUFFERS) return NULL;
  return images[idx].data;
}

void gs_copy_image(int dst_idx, int src_idx) {
  if (dst_idx < 0 || dst_idx >= NUM_BUFFERS || src_idx < 0 || src_idx >= NUM_BUFFERS) return;
  gs_copy(images[dst_idx], images[src_idx]);
}

void gs_blur_image(int dst_idx, int src_idx, int radius) {
  if (dst_idx < 0 || dst_idx >= NUM_BUFFERS || src_idx < 0 || src_idx >= NUM_BUFFERS) return;
  gs_blur(images[dst_idx], images[src_idx], radius);
}

uint8_t gs_otsu_threshold_image(int src_idx) {
  if (src_idx < 0 || src_idx >= NUM_BUFFERS) return 0;
  return gs_otsu_threshold(images[src_idx]);
}

void gs_threshold_image(int img_idx, uint8_t threshold) {
  if (img_idx < 0 || img_idx >= NUM_BUFFERS) return;
  gs_threshold(images[img_idx], threshold);
}

void gs_adaptive_threshold_image(int dst_idx, int src_idx, int block_size) {
  if (dst_idx < 0 || dst_idx >= NUM_BUFFERS || src_idx < 0 || src_idx >= NUM_BUFFERS) return;
  gs_adaptive_threshold(images[dst_idx], images[src_idx], block_size | 1, 2);
}

void gs_erode_image(int dst_idx, int src_idx) {
  if (dst_idx < 0 || dst_idx >= NUM_BUFFERS || src_idx < 0 || src_idx >= NUM_BUFFERS) return;
  gs_erode(images[dst_idx], images[src_idx]);
}

void gs_erode_image_iterations(int dst_idx, int src_idx, int iterations) {
  if (dst_idx < 0 || dst_idx >= NUM_BUFFERS || src_idx < 0 || src_idx >= NUM_BUFFERS) return;

  // First iteration
  gs_erode(images[dst_idx], images[src_idx]);

  // Additional iterations (ping-pong between buffers)
  int temp_idx = (dst_idx + 1) % NUM_BUFFERS;
  if (temp_idx == src_idx) temp_idx = (temp_idx + 1) % NUM_BUFFERS;

  for (int i = 1; i < iterations; i++) {
    if (i % 2 == 1) {
      gs_erode(images[temp_idx], images[dst_idx]);
    } else {
      gs_erode(images[dst_idx], images[temp_idx]);
    }
  }

  // Ensure result is in dst_idx
  if (iterations > 1 && iterations % 2 == 0) { gs_copy(images[dst_idx], images[temp_idx]); }
}

void gs_dilate_image(int dst_idx, int src_idx) {
  if (dst_idx < 0 || dst_idx >= NUM_BUFFERS || src_idx < 0 || src_idx >= NUM_BUFFERS) return;
  gs_dilate(images[dst_idx], images[src_idx]);
}

void gs_dilate_image_iterations(int dst_idx, int src_idx, int iterations) {
  if (dst_idx < 0 || dst_idx >= NUM_BUFFERS || src_idx < 0 || src_idx >= NUM_BUFFERS) return;

  // First iteration
  gs_dilate(images[dst_idx], images[src_idx]);

  // Additional iterations (ping-pong between buffers)
  int temp_idx = (dst_idx + 1) % NUM_BUFFERS;
  if (temp_idx == src_idx) temp_idx = (temp_idx + 1) % NUM_BUFFERS;

  for (int i = 1; i < iterations; i++) {
    if (i % 2 == 1) {
      gs_dilate(images[temp_idx], images[dst_idx]);
    } else {
      gs_dilate(images[dst_idx], images[temp_idx]);
    }
  }

  // Ensure result is in dst_idx
  if (iterations > 1 && iterations % 2 == 0) { gs_copy(images[dst_idx], images[temp_idx]); }
}

void gs_sobel_image(int dst_idx, int src_idx) {
  if (dst_idx < 0 || dst_idx >= NUM_BUFFERS || src_idx < 0 || src_idx >= NUM_BUFFERS) return;
  gs_sobel(images[dst_idx], images[src_idx]);
}

// Blob detection functions
static gs_label labels_buffer[640 * 480];  // Max image size buffer for labels
static struct gs_blob blobs_buffer[200];   // Buffer for blob storage

unsigned gs_detect_blobs(int src_idx, unsigned max_blobs) {
  if (src_idx < 0 || src_idx >= NUM_BUFFERS) return 0;
  if (max_blobs > 200) max_blobs = 200;  // Limit to buffer size

  unsigned size = images[src_idx].w * images[src_idx].h;
  for (unsigned i = 0; i < size; i++) labels_buffer[i] = 0;
  for (unsigned i = 0; i < max_blobs; i++) blobs_buffer[i] = (struct gs_blob){0};

  return gs_blobs(images[src_idx], labels_buffer, blobs_buffer, max_blobs);
}

struct gs_blob* gs_get_blob(unsigned idx) {
  if (idx >= 200) return NULL;
  return &blobs_buffer[idx];
}

gs_label* gs_get_labels_buffer(void) { return labels_buffer; }

// Blob corners buffer
static struct gs_point blob_corners_buffer[4];

void gs_get_blob_corners(unsigned blob_idx) {
  if (blob_idx >= 200) return;
  gs_blob_corners(images[0], labels_buffer, &blobs_buffer[blob_idx], blob_corners_buffer);
}

struct gs_point* gs_get_blob_corner(unsigned corner_idx) {
  if (corner_idx >= 4) return NULL;
  return &blob_corners_buffer[corner_idx];
}

// Contour tracing for largest blob
static struct gs_contour largest_blob_contour;
static uint8_t contour_visited_buffer[640 * 480];

void gs_trace_largest_blob_contour(int src_idx) {
  if (src_idx < 0 || src_idx >= NUM_BUFFERS) return;

  // Find largest blob
  unsigned largest_idx = 0;
  for (unsigned i = 1; i < 200; i++) {
    if (blobs_buffer[i].area > blobs_buffer[largest_idx].area) { largest_idx = i; }
  }

  if (blobs_buffer[largest_idx].area == 0) return;

  // Set up visited buffer
  struct gs_image visited = {images[src_idx].w, images[src_idx].h, contour_visited_buffer};
  for (unsigned i = 0; i < visited.w * visited.h; i++) { visited.data[i] = 0; }

  // Find a starting point on the blob boundary
  largest_blob_contour.start =
      (struct gs_point){blobs_buffer[largest_idx].box.x, blobs_buffer[largest_idx].box.y};

  // Find actual boundary point by scanning the blob's bounding box
  for (unsigned y = blobs_buffer[largest_idx].box.y;
       y < blobs_buffer[largest_idx].box.y + blobs_buffer[largest_idx].box.h; y++) {
    for (unsigned x = blobs_buffer[largest_idx].box.x;
         x < blobs_buffer[largest_idx].box.x + blobs_buffer[largest_idx].box.w; x++) {
      if (x < images[src_idx].w && y < images[src_idx].h &&
          images[src_idx].data[y * images[src_idx].w + x] > 128) {
        largest_blob_contour.start = (struct gs_point){x, y};
        goto found_start;
      }
    }
  }

found_start:
  gs_trace_contour(images[src_idx], visited, &largest_blob_contour);
}

struct gs_contour* gs_get_largest_blob_contour(void) { return &largest_blob_contour; }

// FAST keypoint detection
static struct gs_keypoint keypoints_buffer[500];
static uint8_t scoremap_buffer[640 * 480];

unsigned gs_detect_fast_keypoints(int src_idx, unsigned threshold, unsigned max_keypoints) {
  if (src_idx < 0 || src_idx >= NUM_BUFFERS) return 0;
  if (max_keypoints > 500) max_keypoints = 500;

  struct gs_image scoremap = {images[src_idx].w, images[src_idx].h, scoremap_buffer};
  return gs_fast(images[src_idx], scoremap, keypoints_buffer, max_keypoints, threshold);
}

struct gs_keypoint* gs_get_keypoint(unsigned idx) {
  if (idx >= 500) return NULL;
  return &keypoints_buffer[idx];
}

// ORB feature extraction
static struct gs_keypoint orb_keypoints_buffer[300];
static struct gs_keypoint template_keypoints_buffer[300];
static struct gs_match matches_buffer[200];

unsigned gs_extract_orb_features(int src_idx, unsigned threshold, unsigned max_keypoints) {
  if (src_idx < 0 || src_idx >= NUM_BUFFERS) return 0;
  if (max_keypoints > 300) max_keypoints = 300;

  return gs_orb_extract(images[src_idx], orb_keypoints_buffer, max_keypoints, threshold,
                        scoremap_buffer);
}

struct gs_keypoint* gs_get_orb_keypoint(unsigned idx) {
  if (idx >= 300) return NULL;
  return &orb_keypoints_buffer[idx];
}

void gs_store_template_keypoints(unsigned count) {
  if (count > 300) count = 300;
  for (unsigned i = 0; i < count; i++) { template_keypoints_buffer[i] = orb_keypoints_buffer[i]; }
}

unsigned gs_match_orb_features(unsigned template_count, unsigned scene_count, float max_distance) {
  if (template_count > 300) template_count = 300;
  if (scene_count > 300) scene_count = 300;

  return gs_match_orb(template_keypoints_buffer, template_count, orb_keypoints_buffer, scene_count,
                      matches_buffer, 200, max_distance);
}

struct gs_match* gs_get_match(unsigned idx) {
  if (idx >= 200) return NULL;
  return &matches_buffer[idx];
}

struct gs_keypoint* gs_get_template_keypoint(unsigned idx) {
  if (idx >= 300) return NULL;
  return &template_keypoints_buffer[idx];
}

// Contour detection for largest blob
static struct gs_contour contour_buffer;
static uint8_t visited_buffer[640 * 480];

int gs_detect_largest_blob_contour(int src_idx, unsigned max_blobs) {
  if (src_idx < 0 || src_idx >= NUM_BUFFERS) return 0;

  // First detect blobs
  unsigned num_blobs = gs_detect_blobs(src_idx, max_blobs);
  if (num_blobs == 0) return 0;

  // Find largest blob
  unsigned largest_idx = 0;
  unsigned largest_area = blobs_buffer[0].area;
  for (unsigned i = 1; i < num_blobs; i++) {
    if (blobs_buffer[i].area > largest_area) {
      largest_area = blobs_buffer[i].area;
      largest_idx = i;
    }
  }

  if (largest_area < 100) return 0;  // Skip very small blobs

  // Initialize visited map
  unsigned size = images[src_idx].w * images[src_idx].h;
  for (unsigned i = 0; i < size; i++) visited_buffer[i] = 0;

  struct gs_image visited = {images[src_idx].w, images[src_idx].h, visited_buffer};

  // Find a starting point on the blob boundary
  struct gs_blob* blob = &blobs_buffer[largest_idx];
  contour_buffer.start.x = blob->box.x;
  contour_buffer.start.y = blob->box.y;

  // Find first boundary pixel
  int found = 0;
  for (unsigned y = blob->box.y; y < blob->box.y + blob->box.h && !found; y++) {
    for (unsigned x = blob->box.x; x < blob->box.x + blob->box.w && !found; x++) {
      if (labels_buffer[y * images[src_idx].w + x] == blob->label) {
        contour_buffer.start.x = x;
        contour_buffer.start.y = y;
        found = 1;
      }
    }
  }

  if (!found) return 0;

  // Trace contour
  gs_trace_contour(images[src_idx], visited, &contour_buffer);

  return contour_buffer.length > 0 ? 1 : 0;
}

struct gs_contour* gs_get_contour(void) { return &contour_buffer; }

// LBP Face detection
static uint32_t integral_buffer[640 * 480];
static struct gs_rect faces_buffer[100];

unsigned gs_detect_faces(int src_idx, int min_neighbors) {
  if (src_idx < 0 || src_idx >= NUM_BUFFERS) return 0;
  if (images[src_idx].w * images[src_idx].h > 640 * 480) return 0;

  gs_integral(images[src_idx], integral_buffer);
  return gs_lbp_detect(&frontalface, integral_buffer, images[src_idx].w, images[src_idx].h,
                       faces_buffer, 100, 1.2f, 1.0f, 4.0f, min_neighbors);
}

struct gs_rect* gs_get_face(unsigned idx) {
  if (idx >= 100) return NULL;
  return &faces_buffer[idx];
}

File: examples/wasm/grayskull.js
const video = document.getElementById('video');
const canvas = document.getElementById('canvas');
const ctx = canvas.getContext('2d');
const startButton = document.getElementById('start');
const stopButton = document.getElementById('stop');
const cameraSelect = document.getElementById('cameraSelect');
const addStepButton = document.getElementById('add-step');
const pipelineStepsContainer = document.getElementById('pipeline-steps');

let wasm, memory;
let videoStream;
let animationFrameId;
let imageBuffers = [null, null, null]; // To hold pointers to WASM memory
let overlayCtx;  // For drawing overlays
let templateKeypoints = null;  // For ORB template matching

const operations = {
    "blur": { params: [{ name: "radius", type: "range", default: 3, min: 1, max: 20 }] },
    "sobel": { params: [] },
    "otsu": { params: [] },
    "thresh": { params: [{ name: "threshold", type: "range", default: 128, min: 0, max: 255 }] },
    "adaptive": { params: [{ name: "block_size", type: "range", default: 15, min: 3, max: 51, step: 2 }] },
    "erode": { params: [{ name: "iterations", type: "range", default: 1, min: 1, max: 10 }] },
    "dilate": { params: [{ name: "iterations", type: "range", default: 1, min: 1, max: 10 }] },
    "blobs": { params: [{ name: "max_blobs", type: "range", default: 10, min: 1, max: 100 }] },
    "contour": { params: [] },
    "keypoints": { params: [{ name: "threshold", type: "range", default: 10, min: 5, max: 100 }, { name: "max_points", type: "range", default: 300, min: 10, max: 2000 }] },
    "orb": { params: [{ name: "threshold", type: "range", default: 30, min: 5, max: 500 }, { name: "max_features", type: "range", default: 100, min: 10, max: 300 }] },
};

// --- Grayscale Conversion in JavaScript ---
function rgbaToGray(rgba, gray) {
    for (let i = 0; i < gray.length; i++) {
        const r = rgba[i * 4], g = rgba[i * 4 + 1], b = rgba[i * 4 + 2];
        gray[i] = 0.299 * r + 0.587 * g + 0.114 * b;
    }
}

function grayToRgba(gray, rgba) {
    for (let i = 0; i < gray.length; i++) {
        rgba[i * 4] = rgba[i * 4 + 1] = rgba[i * 4 + 2] = gray[i];
        rgba[i * 4 + 3] = 255; // Alpha
    }
}

// Template capture functionality
function captureTemplate() {
    if (!wasm || !imageBuffers[0]) {
        alert('Camera not running');
        return;
    }

    const width = canvas.width;
    const height = canvas.height;
    const imageSize = width * height;

    // Capture current frame as template
    ctx.drawImage(video, 0, 0, width, height);
    const imageData = ctx.getImageData(0, 0, width, height);
    const gray = new Uint8Array(imageSize);
    rgbaToGray(imageData.data, gray);
    new Uint8Array(memory.buffer, imageBuffers[0], imageSize).set(gray);

    // Extract ORB features from template
    const templateFeatures = wasm.gs_extract_orb_features(0, 20, 200);
    if (templateFeatures > 0) {
        wasm.gs_store_template_keypoints(templateFeatures);
        templateKeypoints = { count: templateFeatures };
        document.getElementById('template-status').textContent = `Template captured: ${templateFeatures} features`;
        console.log(`Template captured with ${templateFeatures} ORB features`);
    } else {
        alert('No features detected in template. Try capturing a more textured area.');
    }
}

async function init() {
    try {
        const importObject = { env: {} };
        // Fallback for WASM loading if streaming fails (MIME type issues)
        let wasmModule;
        try {
            const { instance } = await WebAssembly.instantiateStreaming(fetch('grayskull.wasm'), importObject);
            wasmModule = instance;
        } catch (streamError) {
            console.warn("Streaming failed, trying fallback:", streamError);
            const response = await fetch('grayskull.wasm');
            const bytes = await response.arrayBuffer();
            const { instance } = await WebAssembly.instantiate(bytes, importObject);
            wasmModule = instance;
        }
        wasm = wasmModule.exports;
        memory = wasm.memory;
        console.log("WebAssembly module loaded.");
    } catch (e) {
        console.warn("WASM initialization failed.", e);
        alert("Failed to load grayskull.wasm. Make sure the file is present and the server is running correctly.");
        return;
    }
    await populateCameraList();
    addStepButton.onclick = () => addPipelineStep();
    document.getElementById('capture-template').onclick = captureTemplate;
    addPipelineStep('blur'); // Add a default first step
    startButton.disabled = false;
}

async function populateCameraList() {
    try {
        if (!navigator.mediaDevices || !navigator.mediaDevices.enumerateDevices) {
            throw new Error("Media device enumeration not supported.");
        }

        // Request camera permission first to get device labels
        let permissionStream;
        try {
            permissionStream = await navigator.mediaDevices.getUserMedia({ video: true });
        } catch (permError) {
            console.warn("Could not get camera permission for labeling:", permError);
        }

        const devices = await navigator.mediaDevices.enumerateDevices();
        const videoDevices = devices.filter(d => d.kind === 'videoinput');

        // Stop the permission stream
        if (permissionStream) {
            permissionStream.getTracks().forEach(track => track.stop());
        }

        cameraSelect.innerHTML = '';
        if (videoDevices.length === 0) {
            const option = document.createElement('option');
            option.textContent = 'No cameras found';
            cameraSelect.appendChild(option);
            return;
        }

        videoDevices.forEach((device, i) => {
            const option = document.createElement('option');
            option.value = device.deviceId;
            option.textContent = device.label || `Camera ${i + 1}`;
            cameraSelect.appendChild(option);
        });
    } catch (error) {
        console.error("Could not enumerate video devices:", error);
        // Add a fallback option
        cameraSelect.innerHTML = '<option value="">Default Camera</option>';
    }
}

startButton.onclick = async () => {
    if (videoStream) stopCamera();
    try {
        // More flexible constraints - prefer 320x240 but allow fallback
        const constraints = {
            video: {
                width: { ideal: 320, min: 160, max: 640 },
                height: { ideal: 240, min: 120, max: 480 },
                frameRate: { ideal: 30, max: 60 }
            }
        };

        // Add device constraint if a specific camera is selected
        const selectedDeviceId = cameraSelect.value;
        if (selectedDeviceId) {
            constraints.video.deviceId = { ideal: selectedDeviceId };
        }

        videoStream = await navigator.mediaDevices.getUserMedia(constraints);
        video.srcObject = videoStream;
        await video.play();

        canvas.width = video.videoWidth;
        canvas.height = video.videoHeight;

        console.log(`Video resolution: ${canvas.width}x${canvas.height}`);

        if (wasm) {
            wasm.gs_reset_allocator();
            for (let i = 0; i < imageBuffers.length; i++) {
                wasm.gs_init_image(i, canvas.width, canvas.height);
                imageBuffers[i] = wasm.gs_get_image_data(i);
            }
        }

        animationFrameId = requestAnimationFrame(processFrame);
        startButton.disabled = true;
        stopButton.disabled = false;
    } catch (error) {
        console.error("Failed to start camera:", error);
        alert(`Could not start camera. Error: ${error.message}`);
    }
};

function stopCamera() {
    if (animationFrameId) cancelAnimationFrame(animationFrameId);
    if (videoStream) videoStream.getTracks().forEach(track => track.stop());
    video.srcObject = null;
    animationFrameId = null;
    startButton.disabled = false;
    stopButton.disabled = true;
}

stopButton.onclick = stopCamera;

function addPipelineStep(opName = 'blur') {
    const stepDiv = document.createElement('div');
    stepDiv.className = 'pipeline-step';

    const select = document.createElement('select');
    for (const key in operations) {
        const option = document.createElement('option');
        option.value = key;
        option.textContent = key;
        if (key === opName) option.selected = true;
        select.appendChild(option);
    }
    stepDiv.appendChild(select);

    const controlsContainer = document.createElement('span');
    stepDiv.appendChild(controlsContainer);

    const removeBtn = document.createElement('button');
    removeBtn.textContent = 'Remove';
    removeBtn.onclick = () => stepDiv.remove();
    stepDiv.appendChild(removeBtn);

    select.onchange = () => updateStepControls(controlsContainer, select.value);
    updateStepControls(controlsContainer, opName);

    pipelineStepsContainer.appendChild(stepDiv);
}

function updateStepControls(container, opName) {
    container.innerHTML = '';
    const op = operations[opName];
    if (!op || !op.params) return;

    op.params.forEach(param => {
        const input = document.createElement('input');
        input.type = param.type;
        input.name = param.name;
        if (param.type === 'range') {
            input.min = param.min || 0;
            input.max = param.max || 255;
            input.step = param.step || 1;
            input.value = param.default || 0;
            container.appendChild(input);

            const valueSpan = document.createElement('span');
            valueSpan.textContent = `(${input.value})`;
            input.oninput = () => valueSpan.textContent = `(${input.value})`;
            container.appendChild(valueSpan);
        }
    });
}

function processFrame() {
    if (!videoStream || video.paused || video.ended || !wasm) {
        if (animationFrameId) requestAnimationFrame(processFrame);
        return;
    }

    const width = canvas.width;
    const height = canvas.height;
    const imageSize = width * height;

    // 1. Get video frame and convert to grayscale, placing it in buffer 0
    ctx.drawImage(video, 0, 0, width, height);
    const imageData = ctx.getImageData(0, 0, width, height);
    const gray = new Uint8Array(imageSize);
    rgbaToGray(imageData.data, gray);
    new Uint8Array(memory.buffer, imageBuffers[0], imageSize).set(gray);

    // 2. Execute pipeline
    let readIdx = 0;
    let writeIdx = 1;
    let overlayData = [];  // Store data for overlays

    pipelineStepsContainer.querySelectorAll('.pipeline-step').forEach(stepDiv => {
        const op = stepDiv.querySelector('select').value;
        const inputs = stepDiv.querySelectorAll('input');
        const params = Array.from(inputs).map(i => parseInt(i.value, 10));

        switch (op) {
            case 'blur': wasm.gs_blur_image(writeIdx, readIdx, params[0]); break;
            case 'thresh':
                wasm.gs_copy_image(writeIdx, readIdx);
                wasm.gs_threshold_image(writeIdx, params[0]);
                break;
            case 'adaptive': wasm.gs_adaptive_threshold_image(writeIdx, readIdx, params[0]); break;
            case 'erode': wasm.gs_erode_image_iterations(writeIdx, readIdx, params[0] || 1); break;
            case 'dilate': wasm.gs_dilate_image_iterations(writeIdx, readIdx, params[0] || 1); break;
            case 'sobel': wasm.gs_sobel_image(writeIdx, readIdx); break;
            case 'otsu':
                const threshold = wasm.gs_otsu_threshold_image(readIdx);
                wasm.gs_copy_image(writeIdx, readIdx);
                wasm.gs_threshold_image(writeIdx, threshold);
                break;
            case 'blobs':
                wasm.gs_copy_image(writeIdx, readIdx);
                const numBlobs = wasm.gs_detect_blobs(readIdx, params[0] || 10);
                overlayData.push({ type: 'blobs', count: numBlobs });
                break;
            case 'contour':
                wasm.gs_copy_image(writeIdx, readIdx);
                const hasContour = wasm.gs_detect_largest_blob_contour(readIdx, 50);
                if (hasContour) {
                    overlayData.push({ type: 'contour', hasContour: true });
                }
                break;
            case 'keypoints':
                wasm.gs_copy_image(writeIdx, readIdx);
                const numKeypoints = wasm.gs_detect_fast_keypoints(readIdx, params[0] || 20, params[1] || 100);
                overlayData.push({ type: 'keypoints', count: numKeypoints });
                break;
            case 'orb':
                wasm.gs_copy_image(writeIdx, readIdx);
                const numOrbFeatures = wasm.gs_extract_orb_features(readIdx, params[0] || 20, params[1] || 100);
                overlayData.push({ type: 'orb', count: numOrbFeatures });

                // If we have template keypoints, try to match
                if (templateKeypoints && templateKeypoints.count > 0) {
                    const numMatches = wasm.gs_match_orb_features(templateKeypoints.count, numOrbFeatures, 60.0);
                    overlayData.push({ type: 'matches', count: numMatches });
                }
                break;
        }
        // Swap buffers
        [readIdx, writeIdx] = [writeIdx, (writeIdx + 1) % 3 === readIdx ? (writeIdx + 2) % 3 : (writeIdx + 1) % 3];
    });

    // 3. Display result from the last read buffer
    const resultGray = new Uint8Array(memory.buffer, imageBuffers[readIdx], imageSize);
    const imageDataOut = ctx.createImageData(width, height);
    grayToRgba(resultGray, imageDataOut.data);
    ctx.putImageData(imageDataOut, 0, 0);

    // 4. Draw overlays
    drawOverlays(overlayData, width, height);

    animationFrameId = requestAnimationFrame(processFrame);
}

function drawOverlays(overlayData, width, height) {
    if (!overlayData.length) return;

    // Create overlay canvas for drawing annotations
    ctx.strokeStyle = '#ff0000';
    ctx.fillStyle = '#ff0000';
    ctx.lineWidth = 2;

    overlayData.forEach(overlay => {
        switch (overlay.type) {
            case 'blobs':
                drawBlobs(overlay.count);
                break;
            case 'contour':
                if (overlay.hasContour) {
                    drawContour();
                }
                break;
            case 'keypoints':
                drawKeypoints(overlay.count, '#00ff00');
                break;
            case 'orb':
                drawKeypoints(overlay.count, '#0080ff', true);
                break;
            case 'matches':
                drawMatches(overlay.count);
                break;
        }
    });
}

function drawBlobs(count) {
    ctx.strokeStyle = '#ff0000';
    ctx.lineWidth = 2;

    for (let i = 0; i < count; i++) {
        const blobPtr = wasm.gs_get_blob(i);
        if (!blobPtr) continue;

        // Read blob data from WASM memory
        const blobData = new Uint32Array(memory.buffer, blobPtr, 8); // gs_blob struct
        const area = blobData[1];
        if (area < 50) continue; // Skip very small blobs

        const x = blobData[2];
        const y = blobData[3];
        const w = blobData[4];
        const h = blobData[5];
        const centroidX = blobData[6];
        const centroidY = blobData[7];

        // Draw bounding box (rectangle)
        ctx.strokeStyle = '#ff0000';
        ctx.strokeRect(x, y, w, h);

        // Draw centroid
        ctx.fillStyle = '#ff0000';
        ctx.beginPath();
        ctx.arc(centroidX, centroidY, 3, 0, 2 * Math.PI);
        ctx.fill();

        // Draw area text
        ctx.fillStyle = '#ffffff';
        ctx.fillText(`${area}`, x, y - 5);

        // Draw corners
        const cornersPtr = wasm.gs_get_blob_corners(i);
        if (cornersPtr) {
            ctx.fillStyle = '#00ff00';
            const cornersData = new Uint32Array(memory.buffer, cornersPtr, 8); // 4 points, 2 coords each
            for (let j = 0; j < 4; j++) {
                const cornerX = cornersData[j * 2];
                const cornerY = cornersData[j * 2 + 1];
                ctx.beginPath();
                ctx.arc(cornerX, cornerY, 2, 0, 2 * Math.PI);
                ctx.fill();
            }
        }
    }
}

function drawContour() {
    const contourPtr = wasm.gs_get_contour();
    if (!contourPtr) return;

    // Read contour data
    const contourData = new Uint32Array(memory.buffer, contourPtr, 6); // gs_contour struct
    const startX = contourData[4];
    const startY = contourData[5];
    const length = contourData[6];

    if (length === 0) return;

    // Draw contour as a highlighted outline
    ctx.strokeStyle = '#ffff00';
    ctx.lineWidth = 2;

    // For simplicity, draw a circle at the start point to indicate contour detection
    ctx.beginPath();
    ctx.arc(startX, startY, 8, 0, 2 * Math.PI);
    ctx.stroke();

    // Draw length text
    ctx.fillStyle = '#ffff00';
    ctx.fillText(`Contour: ${length}px`, startX + 10, startY - 10);
}

function drawKeypoints(count, color, withOrientation = false) {
    ctx.strokeStyle = color;
    ctx.fillStyle = color;
    ctx.lineWidth = 1;

    for (let i = 0; i < count; i++) {
        const kpPtr = withOrientation ? wasm.gs_get_orb_keypoint(i) : wasm.gs_get_keypoint(i);
        if (!kpPtr) continue;

        // Read keypoint data (x, y, response, angle)
        const kpData = new Uint32Array(memory.buffer, kpPtr, 3);
        const x = kpData[0];
        const y = kpData[1];
        const response = kpData[2];

        if (response < 10) continue; // Skip weak keypoints

        // Draw circle
        ctx.beginPath();
        ctx.arc(x, y, 3, 0, 2 * Math.PI);
        ctx.stroke();

        // Draw orientation line for ORB features
        if (withOrientation) {
            const angleData = new Float32Array(memory.buffer, kpPtr + 12, 1);
            const angle = angleData[0];
            const length = 8;
            const endX = x + Math.cos(angle) * length;
            const endY = y + Math.sin(angle) * length;

            ctx.beginPath();
            ctx.moveTo(x, y);
            ctx.lineTo(endX, endY);
            ctx.stroke();
        }
    }
}

function drawMatches(count) {
    // Yellow circles show ORB feature matches between the captured template and current scene
    // Lower distance numbers indicate better matches (more similar features)
    // Use "Capture Template" button to set a reference image for matching
    if (!templateKeypoints || count === 0) return;

    ctx.strokeStyle = '#ffff00';
    ctx.lineWidth = 1;

    for (let i = 0; i < count; i++) {
        const matchPtr = wasm.gs_get_match(i);
        if (!matchPtr) continue;

        // Read match data (idx1, idx2, distance)
        const matchData = new Uint32Array(memory.buffer, matchPtr, 3);
        const templateIdx = matchData[0];
        const sceneIdx = matchData[1];
        const distance = matchData[2];

        if (distance > 40) continue; // Skip poor matches

        // Get scene keypoint
        const sceneKpPtr = wasm.gs_get_orb_keypoint(sceneIdx);
        if (!sceneKpPtr) continue;

        const sceneKpData = new Uint32Array(memory.buffer, sceneKpPtr, 2);
        const sceneX = sceneKpData[0];
        const sceneY = sceneKpData[1];

        // Draw match indicator (yellow circle = matched feature)
        ctx.fillStyle = '#ffff00';
        ctx.beginPath();
        ctx.arc(sceneX, sceneY, 5, 0, 2 * Math.PI);
        ctx.fill();

        // Draw match distance (lower = better match)
        ctx.fillStyle = '#000000';
        ctx.fillText(`${distance}`, sceneX + 6, sceneY - 6);
        ctx.fillStyle = '#ffffff';
        ctx.fillText(`${distance}`, sceneX + 6, sceneY - 6);
    }
}

init();